File: packages/sdkwork-utils-python/sdkwork_utils/currency.py
from sdkwork_utils.i18n import format_number_locale
from sdkwork_utils.number import round_number

_KNOWN: dict[str, tuple[int, str]] = {
    "USD": (2, "$"),
    "EUR": (2, "€"),
    "GBP": (2, "£"),
    "CNY": (2, "¥"),
    "JPY": (0, "¥"),
    "KRW": (0, "₩"),
    "HKD": (2, "HK$"),
    "TWD": (2, "NT$"),
    "CHF": (2, "CHF"),
    "CAD": (2, "CA$"),
    "AUD": (2, "A$"),
    "INR": (2, "₹"),
    "BHD": (3, "BHD"),
    "KWD": (3, "KWD"),
}


def _lookup(code: str) -> tuple[int, str] | None:
    normalized = code.strip()
    if len(normalized) != 3 or normalized != normalized.upper() or not normalized.isalpha():
        return None
    return _KNOWN.get(normalized)


def is_currency_code(value: str) -> bool:
    return _lookup(value) is not None


def minor_unit_exponent(code: str) -> int | None:
    meta = _lookup(code)
    return meta[0] if meta else None


def to_minor_units(amount: float, code: str) -> int | None:
    meta = _lookup(code)
    if meta is None:
        return None
    exponent, _ = meta
    factor = 10**exponent
    return int(round_number(amount * factor, 0))


def from_minor_units(minor: int, code: str) -> float | None:
    meta = _lookup(code)
    if meta is None:
        return None
    exponent, _ = meta
    factor = 10**exponent
    return minor / factor


def _suffix_locale(locale: str) -> bool:
    normalized = locale.lower()
    return normalized.startswith(("de", "fr", "it", "es"))


def format_currency(amount: float, code: str, locale: str) -> str | None:
    meta = _lookup(code)
    if meta is None:
        return None
    exponent, symbol = meta
    formatted = format_number_locale(amount, locale, exponent)
    if _suffix_locale(locale):
        return f"{formatted} {symbol}"
    return f"{symbol}{formatted}"

File: packages/sdkwork-utils-typescript/src/currency.ts
import { formatNumberLocale } from "./i18n.js";
import { round } from "./number.js";

const KNOWN: Record<string, { exponent: number; symbol: string }> = {
  USD: { exponent: 2, symbol: '$' },
  EUR: { exponent: 2, symbol: '\u20AC' },
  GBP: { exponent: 2, symbol: '\u00A3' },
  CNY: { exponent: 2, symbol: '\u00A5' },
  JPY: { exponent: 0, symbol: '\u00A5' },
  KRW: { exponent: 0, symbol: '\u20A9' },
  HKD: { exponent: 2, symbol: 'HK$' },
  TWD: { exponent: 2, symbol: 'NT$' },
  CHF: { exponent: 2, symbol: 'CHF' },
  CAD: { exponent: 2, symbol: 'CA$' },
  AUD: { exponent: 2, symbol: 'A$' },
  INR: { exponent: 2, symbol: '\u20B9' },
  BHD: { exponent: 3, symbol: 'BHD' },
  KWD: { exponent: 3, symbol: 'KWD' },
};

function lookup(code: string) {
  const trimmed = code.trim();
  if (!/^[A-Z]{3}$/.test(trimmed)) {
    return undefined;
  }
  return KNOWN[trimmed];
}

export function isCurrencyCode(value: string): boolean {
  return lookup(value) !== undefined;
}

export function minorUnitExponent(code: string): number | null {
  return lookup(code)?.exponent ?? null;
}

export function toMinorUnits(amount: number, code: string): number | null {
  const meta = lookup(code);
  if (!meta) {
    return null;
  }
  const factor = 10 ** meta.exponent;
  return Math.trunc(round(amount * factor, 0));
}

export function fromMinorUnits(minor: number, code: string): number | null {
  const meta = lookup(code);
  if (!meta) {
    return null;
  }
  const factor = 10 ** meta.exponent;
  return minor / factor;
}

function suffixLocale(locale: string): boolean {
  const normalized = locale.toLowerCase();
  return (
    normalized.startsWith("de") ||
    normalized.startsWith("fr") ||
    normalized.startsWith("it") ||
    normalized.startsWith("es")
  );
}

export function formatCurrency(amount: number, code: string, locale: string): string | null {
  const meta = lookup(code);
  if (!meta) {
    return null;
  }
  const formatted = formatNumberLocale(amount, locale, meta.exponent);
  return suffixLocale(locale) ? `${formatted} ${meta.symbol}` : `${meta.symbol}${formatted}`;
}
